### PR TITLE
MRG: Use caching in Github Actions and Azure Pipelines

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
This PR follows https://github.com/mne-tools/mne-python/issues/8685#issuecomment-753985292 and attempts to solve the recent failures related to testing data download.

ToDo:

- [x] Use `$HOME` for `path` variable
- [x] Use caching for Azure too

### CODE REVIEW

**Comment by GuillaumeFavelier** [comment]
Among the first batch of runners, no one detected the cache:

```
Run actions/cache@v2
Cache not found for input keys: macOS-dataset
```
and

```
Run actions/cache@v2
Cache not found for input keys: Linux-dataset
```

Failing skips the `Post Cache dataset` step.


I'll push an empty commit to confirm that they at least restore the cache successfully.

File: .github/workflows/compat_minimal.yml [hunk 1]
```diff
@@ -64,7 +64,14 @@ jobs:
...
         run: ./tools/github_actions_infos.sh
         name: 'Show infos'
+      - uses: actions/cache@v2
+        id: cache-dataset
+        with:
+          path: /home/runner/mne_data
+          key: ${{ runner.os }}-dataset
+        name: 'Cache dataset'
       - shell: bash -el {0}
+        if: steps.cache-dataset.outputs.cache-hit != 'true'
```
Comment (larsoner) [inline .github/workflows/compat_minimal.yml:74]: We should probably always try to download it, it should be a no-op if the files already exist. We just need to make sure that the version-number checking actually works (e.g., if we specify testing dataset version 0.110 but 0.109 is found in version.txt, that the new one gets downloaded)

**Comment by GuillaumeFavelier** [comment]
`compat_old` was the first to finish `Post Cache dataset` with:

```
Post job cleanup.
/bin/tar --posix --use-compress-program zstd -T0 -cf cache.tzst -P -C /home/runner/work/mne-python/mne-python --files-from manifest.txt
Cache saved successfully
```

**Comment by GuillaumeFavelier** [comment]
I see no problem at restoration :ok_hand: :

```
Run actions/cache@v2
Received 79691776 of 1053312637 (7.6%), 75.9 MBs/sec
Received 209715200 of 1053312637 (19.9%), 99.9 MBs/sec
Received 360710144 of 1053312637 (34.2%), 114.6 MBs/sec
Received 486539264 of 1053312637 (46.2%), 115.9 MBs/sec
Received 629145600 of 1053312637 (59.7%), 119.9 MBs/sec
Received 750780416 of 1053312637 (71.3%), 119.2 MBs/sec
Received 864026624 of 1053312637 (82.0%), 117.6 MBs/sec
Received 981467136 of 1053312637 (93.2%), 116.9 MBs/sec
Received 1053312637 of 1053312637 (100.0%), 72.2 MBs/sec
Cache Size: ~1005 MB (1053312637 B)
/bin/tar --use-compress-program zstd -d -xf /home/runner/work/_temp/6a73b959-0e9e-4b55-b734-c336d86abe67/cache.tzst -P -C /home/runner/work/mne-python/mne-python
Cache restored from key: Linux-dataset
```

**Comment by larsoner** [comment]
Do you want to try dealing with testing dataset `version.txt` number mismatches here, or do you want me to open a quick PR for this?

**Comment by GuillaumeFavelier** [comment]
You'll probably be more efficient than me :sweat_smile: . Go ahead if you have the time.

**Comment by larsoner** [comment]
Will do now, could really use something in my wheelhouse to jump-start things over here...

**Comment by larsoner** [comment]
@GuillaumeFavelier if you want to look into it, we should do the same for Azure

https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=10834&view=logs&jobId=f8b184f0-8536-5d07-cb2e-df31a09e5d0f&j=f8b184f0-8536-5d07-cb2e-df31a09e5d0f&t=417aa5f0-1d30-5c85-3b8c-27e749125871

**Comment by larsoner** [comment]
@GuillaumeFavelier you can use something like the following in a one-liner `tools/get_testing_version.sh` script that could be used with each action:
```
$ grep -oP "(?<=testing=')[0-9.]*(?=')" mne/datasets/utils.py
0.112
```
When Dan's #8679 is in progress (probably after my PR then yours) then this line can probably be adjusted fairly easily

**Comment by GuillaumeFavelier** [comment]
Testing the new key based on the testing version is a success!

`Cache testing data`

```
Run actions/cache@v2
Cache not found for input keys: 0.112
```

**Comment by GuillaumeFavelier** [comment]
The Perl extension is not available on MacOS:

```
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
	[-e pattern] [-f file] [--binary-files=value] [--color=when]
	[--context[=num]] [--directories=action] [--label] [--line-buffered]
	[--null] [pattern] [file ...]
Error: Process completed with exit code 2.
```

**Comment by GuillaumeFavelier** [comment]
Also, the cache is different depending on the OS:

On MacOS:

```
Run actions/cache@v2
Cache not found for input keys: 0.112
```

For the others (on Linux):

```
Cache restored from key: 0.112
```

**Comment by GuillaumeFavelier** [comment]
Because there is a `.` in `0.112` it is interpreted as a file instead of a string by Azure...

**Comment by GuillaumeFavelier** [comment]
I don't know what to say :facepalm: 

```
Resolving key:
 - 0-112 [string]
Resolved to: 0-112
ApplicationInsightsTelemetrySender will correlate events with X-TFS-Session a090b4b2-8116-473d-833f-0e69d60908b7
Getting a pipeline cache artifact with one of the following fingerprints:
Fingerprint: `0-112`
There is a cache miss.
ApplicationInsightsTelemetrySender correlated 1 events with X-TFS-Session a090b4b2-8116-473d-833f-0e69d60908b7
Finishing: Cache testing data
```

**Comment by GuillaumeFavelier** [comment]
The jobs on Windows detected the key:

```
Resolving key:
 - 0-112 [string]
Resolved to: 0-112
ApplicationInsightsTelemetrySender will correlate events with X-TFS-Session 77543406-ba53-47b0-a210-8120a5abf696
Getting a pipeline cache artifact with one of the following fingerprints:
Fingerprint: `0-112`
There is a cache hit: `0-112`
```
And restored the cache correctly

```
Folders: 103
Files: 738
Size:       1613467701
Compressed: 452608
Process exit code: 0
Cache restored.
```

But the data are downloaded anyway because the path is different on Windows.

**Comment by larsoner** [comment]
Hah! 44 sec to download the data on `master`-like runs:

https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=10873&view=logs&j=fa98c7b3-575e-55b6-6946-22dea6f6daf8

And 1m 11 sec to get it from cache:

https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=10886&view=logs&jobId=8d04c17d-3f6b-5313-febd-debdd8b57f4c&j=8d04c17d-3f6b-5313-febd-debdd8b57f4c&t=76e98127-12cf-5f50-cd68-526e1d36628d

So we actually get a slowdown from this on Azure at least (!) but at least it should be more reliable.

**Comment by GuillaumeFavelier** [comment]
Hm... But for some reason it downloads more no?

```
Expected size to be downloaded: 1,614.1 MB
```

Compared to:

```
Downloading archive mne-testing-data-0.112.tar.gz to C:\Users\VssAdministrator\mne_data
Downloading https://codeload.github.com/mne-tools/mne-testing-data/tar.gz/0.112 (1000.0 MB)
```

**Comment by larsoner** [comment]
> Hm... But for some reason it downloads more no?

That's probably the uncompressed size, vs the download which is compressed.

**Comment by GuillaumeFavelier** [comment]
Unrelated failure on `linux / conda / py3.8`

<details>

```
mne/viz/tests/test_3d.py::test_plot_source_estimates[testing_data-testing_data-pyvista-surface-vector] PASSED [  6%]
mne/viz/tests/test_3d.py::test_plot_source_estimates[testing_data-testing_data-pyvista-surface-None] PASSED [  6%]
mne/viz/tests/test_3d.py::test_plot_source_estimates[testing_data-testing_data-pyvista-volume-vector] PASSED [  6%]
mne/viz/tests/test_3d.py::test_plot_source_estimates[testing_data-testing_data-pyvista-volume-None] PASSED [  6%]
mne/viz/tests/test_3d.py::test_plot_source_estimates[testing_data-testing_data-pyvista-mixed-vector] PASSED [  6%]
Fatal Python error: Segmentation fault

Thread 0x00007fa9133ad700 (most recent call first):
  File "/usr/share/miniconda/envs/mne/lib/python3.8/threading.py", line 306 in wait
  File "/usr/share/miniconda/envs/mne/lib/python3.8/threading.py", line 558 in wait
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/tqdm/_monitor.py", line 59 in run
  File "/usr/share/miniconda/envs/mne/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/share/miniconda/envs/mne/lib/python3.8/threading.py", line 890 in _bootstrap

Current thread 0x00007fa953cb6740 (most recent call first):
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pyvistaqt/plotting.py", line 696 in set_icon
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pyvistaqt/plotting.py", line 597 in __init__
  File "/home/runner/work/mne-python/mne-python/mne/viz/backends/_pyvista.py", line 95 in build
  File "/home/runner/work/mne-python/mne-python/mne/viz/backends/_pyvista.py", line 199 in __init__
  File "/home/runner/work/mne-python/mne-python/mne/viz/backends/renderer.py", line 38 in _get_renderer
  File "/home/runner/work/mne-python/mne-python/mne/viz/_brain/_brain.py", line 427 in __init__
  File "/home/runner/work/mne-python/mne-python/mne/viz/_3d.py", line 1904 in _plot_stc
  File "/home/runner/work/mne-python/mne-python/mne/viz/_3d.py", line 1824 in plot_source_estimates
  File "<decorator-gen-142>", line 22 in plot_source_estimates
  File "/home/runner/work/mne-python/mne-python/mne/source_estimate.py", line 653 in plot
  File "/home/runner/work/mne-python/mne-python/mne/viz/tests/test_3d.py", line 676 in test_plot_source_estimates
  File "/home/runner/work/mne-python/mne-python/mne/utils/_testing.py", line 251 in dec
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/python.py", line 183 in pytest_pyfunc_call
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/python.py", line 1641 in runtest
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 162 in pytest_runtest_call
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 255 in <lambda>
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 311 in from_call
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 254 in call_runtest_hook
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 215 in call_and_report
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 126 in runtestprotocol
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/main.py", line 348 in pytest_runtestloop
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/main.py", line 323 in _main
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/main.py", line 269 in wrap_session
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/config/__init__.py", line 162 in main
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/config/__init__.py", line 185 in console_main
  File "/usr/share/miniconda/envs/mne/bin/pytest", line 8 in <module>
./tools/github_actions_test.sh: line 22:  4547 Segmentation fault      (core dumped) pytest -m "${CONDITION}" --tb=short --cov=mne --cov-report xml -vv ${USE_DIRS}
mne/viz/tests/test_3d.py::test_plot_source_estimates[testing_data-testing_data-pyvista-mixed-None] 
Error: Process completed with exit code 139.
```

</details>

Maybe one last empty commit for testing but everything is ok for me. Is there something missing @larsoner ?

**Comment by larsoner** [comment]
> Maybe one last empty commit for testing but everything is ok for me. Is there something missing @larsoner ?

I don't think we need the empty commit, I'll just restart the segfaulted GitHub actions build, and it looks like everything else should be okay. Will merge once CIs are happy

**Comment by larsoner** [comment]
Thanks @GuillaumeFavelier !

**Comment by GuillaumeFavelier** [comment]
Something interesting about caching:

>The cache is scoped to the key and branch. The default branch cache is available to other branches.

Reference: https://github.com/actions/cache#cache-scopes

**Pull request merged**